#1012 - Fixed violations detected by the qulice 0.15.2 - puzzle 961-58fc8a23.

danff opened the pull request:
Fixes for #1012 - solved puzzle 961-58fc8a23 in netbout-web/pom.xml:747-752
- Removed excludes from pom.xml in netbout-web and netbout-client;
- Made static methods violating NonStaticMethodCheck checkstyle check;
- Refactor some classes to avoid PMD.ConstructorOnlyInitializesOrCallOtherConstructors violations;
- Refactor some classes to avoid PMD.ConstructorShouldDoInitialization violations.
- Added puzzle with excludes for classes violating the new rules added by the upgrade to the qulice 0.16.2 version.

karato commented:
@danff Many thanks, I will find someone to review it before we merge

karato commented:
@mkordas review it

mkordas commented:
@danff I'm on it

mkordas commented on netbout-client/src/main/java/com/netbout/mock/H2Sql.java:
```diff
@@ -48,6 +48,7 @@
 @ToString
 @Loggable(Loggable.DEBUG)
 @EqualsAndHashCode(of = "file")
+@SuppressWarnings("PMD.ConstructorOnlyInitializesOrCallOtherConstructors")
```
@danff do you think it is false positive here?

danff replied:
@mkordas There is a commit which updated netbout/pom.xml with qulice version 0.16.2.

mkordas replied:
@danff so let's fix them

danff replied:
@mkordas Sorry, they are not introduced in 0.16.2, but in 0.15.2

mkordas commented on netbout-client/src/test/java/com/netbout/client/NbRule.java:
```diff
@@ -45,6 +45,7 @@
     /**
      * Get user.
      * @return User
+     * @checkstyle NonStaticMethodCheck (10 lines)
```
@danff same question, are you suppressing because this is false positive?

danff replied:
@mkordas I am not sure the method is supposed to be static. If yes, no problem to make it static then.

mkordas replied:
@danff if method does not refer to any instance field or method then it should be static

mkordas commented on netbout-web/src/main/java/com/netbout/cached/CdMessages.java:
```diff
@@ -146,6 +146,7 @@ public long unread() throws IOException {
+         * @checkstyle NonStaticMethodCheck (5 lines)
          */
         @Cacheable.FlushBefore
         public void touch() {
```
@danff can't it be static?

danff replied:
@mkordas I am not sure the method is supposed to be static. If yes, no problem to make it static then.

mkordas replied:
@danff why do you think so?

danff replied:
@mkordas It is the Flag.touch() method. And it has an annotation Cacheable.FlushBefore. From the annotations javadoc "Identifies a method that should flush all cached entities of **this** class/object, before being executed." I think is should no be static but an instance method.

mkordas replied:
@danff all right, thanks for explanation

mkordas commented on netbout-web/src/main/java/com/netbout/dynamo/DyBase.java:
```diff
@@ -50,6 +50,7 @@
 @Loggable(Loggable.DEBUG)
 @ToString(of = "region")
 @EqualsAndHashCode(of = "region")
+@SuppressWarnings("PMD.ConstructorOnlyInitializesOrCallOtherConstructors")
```
@danff we need to fix this constructor

danff replied:
@mkordas The same violation introduced by the qulice 0.16.2.

mkordas replied:
@danff OK, but how person solving next puzzle will know that this suppression should also be fixed?

danff replied:
@mkordas No it is not 0.16.2. It is related to 0.15.2, so I will try to fix it and remove the SuppressWarnings annotation.

mkordas commented on netbout-web/src/main/java/com/netbout/misc/Ports.java:
```diff
@@ -53,13 +53,14 @@
      * @throws IOException If fails
+     * @checkstyle NonStaticMethodCheck (24 lines)
      */
     public int allocate() throws IOException {
```
@danff should be declared `static`

danff replied:
@mkordas Ok.

mkordas commented on netbout-web/src/main/java/com/netbout/misc/Ports.java:
```diff
@@ -78,6 +79,7 @@ public int allocate() throws IOException {
     /**
      * Release it.
      * @param port Port
+     * @checkstyle NonStaticMethodCheck (4 lines)
```
@danff same here, we need to make it static if it not uses any fields

mkordas commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -58,6 +58,7 @@
  *  this task citing a NullPointerException (See qulice issue #608 raised for
  *  the same) and then remove TkStart.java from qulice exceptions.
  */
+@SuppressWarnings("PMD.ConstructorShouldDoInitialization")
```
@danff we need to fix this constructor instead, it's our principle

danff replied:
@mkordas The violation is from qulice 0.16.2 and is related to the transient field **Date checked**, which is updated **act()** method. No sure it is possible to be only initialized in constructor as it has a specific logic during the **act()** method.

mkordas replied:
@danff it should be possible to refactor, please read http://www.yegor256.com/2015/05/07/ctors-must-be-code-free.html

mkordas commented on netbout-client/src/main/java/com/netbout/client/retry/ReAlias.java:
```diff
@@ -70,7 +70,7 @@ public ReAlias(final Alias orgn) {
         verbose = false, attempts = Tv.TWENTY,
         delay = Tv.FIVE, unit = TimeUnit.SECONDS
-    )
+        )
```
@danff why you changed it? Previous indentation looked better

danff replied:
@mkordas Yes. It was looking better, and sure it is better, but again, violation of qulice from 0.15.2 and also with 0.16.2.

mkordas replied:
@danff I think it was fixed in most recent Qulice, I've raised PR for that

mkordas replied:
@danff please see https://github.com/teamed/qulice/issues/411 - I think it was fixed

mkordas commented on netbout-web/pom.xml:
```diff
@@ -744,24 +744,15 @@
-                                 (lines 192 - 207) related to this puzzle.
-                                 They also need to be fixed in order to solve this puzzle.
+                                @todo #1012:30min These excludes were added because
+                                 of the upgrade to qulice 0.16.2. The new version of qulice
```
@danff where we did upgrade to `0.16.2`?

danff replied:
@mkordas There is a commit which updates netbout/pom.xml to qulice 0.16.2. Actually all violations with qulice 0.15.2 (the ticket is addressed to) are fixed (exception some NonStaticMethod). The puzzle is only for violations which are related to 0.16.2

mkordas replied:
@danff OK, I see

mkordas commented on netbout-web/src/test/java/com/netbout/rest/bout/XeMessageTest.java:
```diff
@@ -70,6 +69,7 @@ public String author() {
                     }
                 }
             ).toXembly()
+        // @checkstyle IndentationCheck (1 line)
```
@danff why this can't be fixed?

danff replied:
@mkordas That is not a workaround. qulice requires to insert 4 spaces between the next closing bracket and the dot.

mkordas replied:
@danff can't we just move `.xml()` to a new line?

danff replied:
@mkordas Same problem - '.' have incorrect indentation level 8, expected level should be 12. (IndentationCheck)

danff replied:
@mkordas See two options, one put 12 char indentation to expect violations of future qulice version upgrades or put the class in the puzzle.

mkordas replied:
@danff OK, let's keep the suppression

mkordas commented:
@danff 
> * Added some PMD suppress warnings when field initialization outside constructor happens;
* Added some PMD suppress warnings when constructor has some more logic then field only initializations.

Danff, the purpose of this issue is to fix things, not to suppress checks...

mkordas commented on netbout-client/src/main/java/com/netbout/client/RtBoutIterator.java:
```diff
@@ -53,6 +53,7 @@
 @ToString
 @Loggable(Loggable.DEBUG)
 @EqualsAndHashCode(of = { "request", "bouts" })
+@SuppressWarnings("PMD.ConstructorShouldDoInitialization")
```
@danff I don't understand your suppression here... What's preventing `bouts` or `more` to be initialized in constructor?

mkordas commented on netbout-client/src/main/java/com/netbout/client/RtMessageIterator.java:
```diff
@@ -56,6 +56,7 @@
 @ToString
 @Loggable(Loggable.DEBUG)
 @EqualsAndHashCode(of = { "request", "messages" })
+@SuppressWarnings("PMD.ConstructorShouldDoInitialization")
```
@danff just initialize `messages` in ctor

mkordas commented on netbout-client/src/main/java/com/netbout/client/retry/ReAttachment.java:
```diff
@@ -69,7 +69,7 @@ public ReAttachment(final Attachment orgn) {
         verbose = false, attempts = Tv.TWENTY,
         delay = Tv.FIVE, unit = TimeUnit.SECONDS
-    )
+        )
```
@danff please revert those changes, indentation was correct

danff replied:
@mkordas qulice does not agree. So should it be added to the puzzle?

mkordas replied:
@danff yes, puzzle is a better option I think

mkordas commented on netbout-web/pom.xml:
```diff
@@ -769,14 +760,10 @@
                                 <exclude>checkstyle:/src/test/java/com/netbout/rest/bout/TkBoutITCase.java</exclude>
-                                <exclude>checkstyle:/src/test/java/com/netbout/rest/bout/TkDeleteITCase.java</exclude>
-                                <exclude>checkstyle:/src/test/java/com/netbout/rest/bout/XeMessageTest.java</exclude>
+                                <exclude>checkstyle:/src/test/java/com/netbout/rest/bout/TkBoutITCase.java</exclude>
```
@danff why some excludes were added while Qulice was not updated?

mkordas replied:
@danff OK, I see now, Qulice was updated in different PR

mkordas commented:
@danff can you please explain the below? why they are needed?
> Added //checkstyle NonStaticMethodCheck on some classes/methods where needed;

mkordas commented:
@danff please see several comments above

mkordas commented:
@danff please also update PR title with more meaningful information about contents of this PR

mkordas commented:
@danff please see my responses

danff commented:
@mkordas Removed all (but one) checkstyle NonMethodCheck and made methods static. Refactor the code and removed the PMD warnings. Updated the description of the PR.

danff commented:
@mkordas The only NonStaticMethodCheck which I have not removed is for the CdMessages.touch() method beause of the Cacheable.FlushBefore annotation.

danff commented:
@mkordas Could you review the new commit?

mkordas commented:
@danff I'm on it again, sorry for delay

mkordas commented on netbout-web/src/main/java/com/netbout/misc/Ports.java:
```diff
@@ -50,16 +50,23 @@
         new ConcurrentSkipListSet<Integer>();
 
     /**
+     * Hide the default no args constructor from the utility class.
```
@danff please see http://www.yegor256.com/2014/05/05/oop-alternative-to-utility-classes.html#utility-classes-are-evil and watch https://www.youtube.com/watch?v=psrp3TtaYYI, we cannot leave it as utility class, but I'm OK for a puzzle

mkordas commented on netbout-web/src/main/java/com/netbout/dynamo/DyBase.java:
```diff
@@ -61,6 +61,24 @@
+     * Initialize a region instance.
+     * @return The initialized region instance
+     */
+    private static Region initialize() {
```
@danff please read http://www.yegor256.com/images/books/elegant-objects/seven-pages.pdf - this method should be named like a noun

mkordas commented on netbout-client/src/main/java/com/netbout/mock/H2Sql.java:
```diff
@@ -60,10 +60,24 @@
+     * @return The name of the JdbcDataSource file name
+     * @throws IOException If fails
+     */
+    private static String initialize() throws IOException {
```
@danff same here - method should be named like a noun

mkordas commented:
@danff please see comments above

mkordas commented:
@danff 
> Removed all (but one) checkstyle NonMethodCheck and made methods static. Refactor the code and removed the PMD warnings. Updated the description of the PR.

thanks for that, good job!

mkordas commented:
@danff can you also adjust PR title to say something meaningful like `#1012 Fix violations detected by Qulice 0.15.2`?

mkordas commented:
@danff ping

mkordas commented:
@danff do you need help here?

mkordas commented:
@danff do you plan to finish this one?

mkordas commented:
@danff ping

mkordas commented:
@danff are you around?

danff commented:
@mkordas I am back. Sorry for delay.

mkordas commented on netbout-web/src/main/java/com/netbout/misc/Ports.java:
```diff
@@ -79,7 +86,7 @@ public int allocate() throws IOException {
      * @param port Port
      */
-    public void release(final int port) {
+    public static void release(final int port) {
```
@danff we shouldn't have any `public static` methods in our code

danff replied:
@mkordas It is the Ports class (utility?) and we have a puzzle to refactor it.

mkordas replied:
@danff right, thanks

mkordas commented on netbout-web/src/main/java/com/netbout/misc/Ports.java:
```diff
@@ -50,16 +50,23 @@
+     * Hide the default no args constructor from the utility class.
+     */
+    private Ports() {
+        super();
```
@danff do we need to call `super()` here?

danff commented:
@mkordas Fixed indentation for RetryOnFailure annotation in netbout-client retry classes. Added puzzle for Ports utility class in netbout-web. Renamed two methods as requested. Leave the ugly indentation in XeMessageTest class to expect violations in new versions of qulice.

mkordas commented on netbout-client/src/main/java/com/netbout/client/retry/ReBout.java:
```diff
@@ -143,7 +143,7 @@ public void subscribe(final boolean subs) throws IOException {
         verbose = false, attempts = Tv.TWENTY,
         delay = Tv.FIVE, unit = TimeUnit.SECONDS
-    )
+        )
```
@danff can we revert that?

mkordas commented on netbout-client/src/test/java/com/netbout/client/NbRule.java:
```diff
@@ -46,7 +46,7 @@
      * @return User
      */
-    public User get() {
+    public static User get() {
```
@danff we can't have any `public static` method

danff replied:
@mkordas Either we have static method, either NonStaticMethodCheck violation. And it is used for tests..

mkordas replied:
@danff but rule is rule, we need at least puzzle for that

danff replied:
@mkordas Ok. I'll do a puzzle for it.

mkordas replied:
@danff looks good

mkordas commented on netbout-web/src/main/java/com/netbout/misc/Ports.java:
```diff
@@ -39,6 +39,9 @@
  * @author Yegor Bugayenko (yegor@teamed.io)
  * @version $Id$
  * @since 2.14.17
+ * @todo #1012:30min This class should not be left as utility one. All methods
```
@danff good puzzle :+1:

mkordas commented:
@danff please see additional comments

danff commented:
@mkordas Removed super() call from Ports. Fixed indentation in ReBout. About static methods can not make not public as the methods are called from other code, and not static bacause of the NonStaticMethodCheck violations.

mkordas commented:
@danff looks good, we just need take care about one more `public static` method

mkordas commented:
@danff ping

mkordas commented:
@danff can you continue here?

mkordas commented:
@danff ping

danff commented:
@mkordas Created a puzzle for the public static method

mkordas commented:
@danff good work!

mkordas commented:
@rultor merge

rultor commented:
> @rultor merge

@mkordas Thanks for your request. @dmzaytsev Please confirm this.

dmzaytsev commented:
@rultor merge

rultor commented:
> @rultor merge

@dmzaytsev OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/7381-190527324)

rultor commented:
> @rultor merge

@dmzaytsev Done! FYI, the full log is [here](http://www.rultor.com/t/7381-190527324) (took me 14min)

dmzaytsev commented:
@rultor deploy

rultor commented:
> @rultor deploy

@dmzaytsev OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/7381-190634587)

rultor commented:
> @rultor deploy

@dmzaytsev Done! FYI, the full log is [here](http://www.rultor.com/t/7381-190634587) (took me 9min)

karato commented:
@ypshenychka review this ticket please, for compliance with our [quality rules](http://at.teamed.io/qa.html)

karato commented:
@rultor deploy

rultor commented:
> @rultor deploy

@karato OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/7381-191175384)

ypshenychka commented:
@karato Quality is good.

rultor commented:
> @rultor deploy

@karato Done! FYI, the full log is [here](http://www.rultor.com/t/7381-191175384) (took me 15min)

karato commented:
> @karato Quality is good.

@ypshenychka thanks!

karato commented:
@mkordas just added 10 mins to @ypshenychka (for QA), payment ID is `79182169`; 10 mins added to @dmzaytsev account (our architect), payment ID is `79182187`, since quality is good here; thanks, I just added **1 hour and 15 mins** to your account, payment `56d6f1c6faa9857232000293`, 276 hours and 58 mins spent; you're getting extra minutes here (c=60); +75 added to your rating, at the moment it is: [+7890](http://www.netbout.com/b/36080?open=rating)
